src/tests/unit_x509.cpp

```diff
@@ -1026,6 +1032,28 @@ Test::Result test_usage(const Botan::Private_Key& ca_key,
    // cert allows every usage
    result.confirm("key usage digitalSignature allowed", no_usage_cert.allowed_usage(Key_Constraints::DIGITAL_SIGNATURE));
    result.confirm("key usage cRLSign allowed", no_usage_cert.allowed_usage(Key_Constraints::CRL_SIGN));
+   result.confirm("key usage encryption allowed", no_usage_cert.allowed_usage(Usage_Type::ENCRYPTION));
```

Comment:
I find this "special case" questionable. Except for RSA these certificates theoretically "allow" everything (including encryption) but practically don't support encryption due to their asymmetric algorithm. Should `allowed_usage` still "claim" encryption?

The creation of a certificate with such an algorithm will throw when specifying either `KEY_ENCIPHERMENT` or `DATA_ENCIPHERMENT` specifically.

Reply:
This is because allowed_usage only looks at the KeyUsage extension which if missing (or wrong) we have no guidance and accept anything, eg a DH cert will be marked as ok for digital signature.

To fix this we must deserialize the public key when parsing and compute the possible constraints. This adds some overhead to cert parsing, but also has some advantages - we can add a new function to `X509_Certificate` that returns a const& to the `Public_Key` so other functions which need to access the public key eg during validation can avoid the extra deserialization step as done in `load_subject_public_key`

Reply:
I see. Thanks for the detailed rationale.
Hence, I understand that the behaviour as sketched out in this test is somewhat expected and "desired", for the time being, right?

Reply:
Yes, expected at least. Probably not really desirable but at least the test does confirm the current behavior. I'd be happy with a PR changing this so we verify or mask the certificate constraints against what the key can actually do, but doesn't need to happen here.